feat: :sparkles: add `write_resource_properties()`

martonvago opened the pull request:
## Description

This adds `write_resource_properties()`.

Closes #757 

<!-- Select quick/in-depth as necessary -->
This PR needs an in-depth review.

## Checklist

- [x] Added or updated tests
- [x] Tests passed locally
- [x] Linted and formatted code
- [x] Build passed locally
- [x] Updated documentation

martonvago committed:
feat: :sparkles: add sensible default values for package properties

martonvago committed:
refactor: :recycle: generalise InvalidResourcePropertiesError to InvalidPropertiesError

martonvago committed:
feat: :sparkles: add verify_package_properties

lwjohnst86 committed:
Merge branch 'main' into feat/verify-package-properties

martonvago committed:
Merge branch 'main' into feat/verify-package-properties

martonvago committed:
refactor: :recycle: change args for InvalidPropertiesError

martonvago committed:
feat: :sparkles: add verify functions for package properties

lwjohnst86 committed:
Merge branch 'main' into feat/verify-package-properties

martonvago committed:
feat: :sparkles: add default values to ResourceProperties

martonvago committed:
refactor: :recycle: rename file invalid_properties_error > not_properties_error

martonvago committed:
refactor: :recycle: rename InvalidPropertiesError > NotPropertiesError

martonvago committed:
refactor: :recycle: generalise verification functions further

martonvago committed:
feat: :sparkles: add resources to required package properties

martonvago committed:
docs: :memo: change wording in docstring

martonvago committed:
refactor: :recycle: only raise errors for given metadata when verifying

signekb committed:
Merge branch 'main' into feat/verify-package-properties

martonvago committed:
feat: :sparkles: make resource property verification analogous to package prop. ver.

martonvago committed:
Merge branch 'main' into feat/resource-default-values

martonvago committed:
Merge branch 'feat/verify-package-properties' into feat/update-verify-resource-properties

martonvago committed:
apply suggestions from code review

Co-authored-by: Signe Kirk Brødbæk <40836345+signekb@users.noreply.github.com>

martonvago committed:
refactor: :recycle: rename verify function files

martonvago committed:
refactor: :recycle: rename verify functions

martonvago committed:
refactor: :recycle: rename variable

martonvago committed:
refactor: :recycle: remove renamed test files

martonvago committed:
Merge branch 'feat/verify-package-properties' into feat/update-verify-resource-properties

martonvago committed:
refactor: :recycle: update function names after rename

martonvago committed:
Merge branch 'main' into feat/verify-package-properties

martonvago committed:
Merge branch 'feat/verify-package-properties' into feat/update-verify-resource-properties

martonvago committed:
test: :white_check_mark: test verify_properties_are_well_formed for resource properties

martonvago committed:
test: :white_check_mark: change test name

martonvago committed:
feat: :sparkles: add write_resource_properties

martonvago commented on sprout/core/write_resource_properties.py:
```diff
@@ -0,0 +1,71 @@
+    return write_json(package_properties, path)
+
+
+def find_resource(package_properties: dict, resource_id: int) -> dict | None:
```
I left these helpers here for now, but we can move them to separate files if we end up using them somewhere else as well.

martonvago committed:
feat: :sparkles: uncomment function in init file

lwjohnst86 committed:
Merge branch 'main' into feat/write-resource-properties

signekb commented on tests/core/test_write_resource_properties.py:
```diff
@@ -0,0 +1,155 @@
+        write_resource_properties(file_path, resource_properties_1.asdict)
+
+
+def test_throws_error_if_resource_properties_incorrect(package_properties_path):
```
```suggestion
def test_throws_error_if_resource_properties_are_incorrect(package_properties_path):
```
? :)

signekb commented on tests/core/test_write_resource_properties.py:
```diff
@@ -0,0 +1,155 @@
+        write_resource_properties(package_properties_path, resource_properties_1.asdict)
+
+
+def test_throws_error_if_package_properties_incorrect(tmp_path, resource_properties_1):
```
```suggestion
def test_throws_error_if_package_properties_are_incorrect(tmp_path, resource_properties_1):
```

signekb commented on tests/core/test_write_resource_properties.py:
```diff
@@ -0,0 +1,155 @@
+        write_resource_properties(package_properties_path, {})
+
+
+def test_throws_error_if_data_path_malformed(
```
Why a test on the path attribute specifically? :)

martonvago replied:
Because if the resource properties object doesn't have an ID in its data path, we cannot compare the IDs of existing resources to this ID.

signekb reviewed (changes requested):
Very nice ⭐ I have commented on some minor things.

martonvago commented on sprout/core/write_resource_properties.py:
```diff
@@ -0,0 +1,71 @@
+    verify_resource_properties(resource_properties)
+
+    package_properties = read_json(path)
+    verify_package_properties(package_properties)
```
That's a good point!
I think the reason I thought it could make sense to check resource properties was that if their data `path` field is somehow broken then we can't read their ID and compare it with the ID of the incoming resource properties.

signekb replied:
Ah, also a very good point! Could we make a check specifically for the `path` field, since this field is especially important?

martonvago replied:
Oh yeah, I think I added a `verify_data_path` function at some point, so we could just use that -- I'll add it in!

martonvago replied:
Actually, I'll make a new PR for that to keep this simpler (because I'm moving `verify_data_path` into its own file.

signekb replied:
Sounds good 🎉

martonvago commented on sprout/core/write_resource_properties.py:
```diff
@@ -0,0 +1,71 @@
+
+    Raises:
+        FileNotFound: If the `datapackage.json` file doesn't exist.
+        NotPropertiesError: If the resource or package properties are not correct.
```
My only problem with this is that we also check whether all our required properties are present on the objects, which is a stricter check than the data package spec alone. And for resources we also check that their data path is well formed.

signekb replied:
True. I just wanted to make it more explicit what "not correct" means.

martonvago replied:
Okay, I added this + the incomplete bit 👍

martonvago commented on sprout/core/write_resource_properties.py:
```diff
@@ -0,0 +1,71 @@
+    package_properties = read_json(path)
+    verify_package_properties(package_properties)
+
+    resource_id = get_resource_id(resource_properties)
```
Don't mind either way!

martonvago committed:
apply suggestions from code review

Co-authored-by: Signe Kirk Brødbæk <40836345+signekb@users.noreply.github.com>

martonvago committed:
refactor: :recycle: review markups

martonvago committed:
docs: :memo: update docstring

lwjohnst86 commented on sprout/core/write_resource_properties.py:
```diff
@@ -0,0 +1,71 @@
+    resource_id = get_resource_id_from_properties(resource_properties)
+    current_resource = get_resource_properties(package_properties, resource_id)
+    if current_resource:
+        current_resource.update(resource_properties)
```
This might have a similar issue as #808. Will this need to be updated?

martonvago replied:
Yeah, I think we have to decide how exactly we want these updates to behave!

lwjohnst86 reviewed (commented):
Very nice :grin: I only have one comment.

martonvago committed:
feat: ✨ verify data path before extracting ID (#807)

## Description

This PR makes sure that every time we extract the ID from the data path
of a resource, we verify that it is well formed
([discussion](https://github.com/seedcase-project/seedcase-sprout/pull/784#discussion_r1801271054)).

Related to #784

<!-- Select quick/in-depth as necessary -->
This PR needs an in-depth review.

## Checklist

- [x] Added or updated tests
- [x] Tests passed locally
- [x] Linted and formatted code
- [x] Build passed locally
- [x] Updated documentation

---------

Co-authored-by: Signe Kirk Brødbæk <40836345+signekb@users.noreply.github.com>
